test/functional/mining_timewarp_fork.py

```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test getblocktemplate complies with potential future timewarp-fixing
+   softforks (modulo 600 second nTime decrease).
+
+   We first mine a chain up to the difficulty-adjustment block and set
+   the last block's timestamp 2 hours in the future, then check the times
+   returned by getblocktemplate. Note that we cannot check the actual
+   retarget behavior as difficulty adjustments do not occur on regtest
+   (though it still technically has a difficulty adjustment interval)."""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    connect_nodes_bi,
+)
+
+
+class TimewarpMiningTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 2
+        self.setup_clean_chain = True
+
+    def run_test(self):
+        self.log.info('Create some old blocks')
+        block_1 = self.nodes[0].generate(1)[0]
+        for i in range(2013):
```

Comment:
Nit: `for _ in range(2013)` is more idiomatic when the index is unused :-)

Reply:
Any reason not to generate all blocks with a singe call here: `self.nodes[0].generate(nblocks=2013)`?